### PR TITLE
[Feature] Support for pdf parsing

### PR DESCRIPTION
**Pull request by Imipenem**
**PR Checklist**

-   [x] This comment contains a description of changes (with reason)
-   [x] Referenced issue is linked
-   [x] If you've fixed a bug or added code that should be tested, add tests!

**Description of changes**

- added setup to parse and cache single pdf's

- added columns_obs_only and index_col support for pdf

- added caching for pdf file 

- multiple pdf file parsing won't be supported for now

- fixed a bug, that caused ehrapy Encoder to fail, when there were only encoded columns after encoding

- fixed caching when used with parsing pdf's

- fixed index column setting for pdf parsing

- refactored `read()` to be more readable

- added `**kwargs` to `read()` to make it more flexible (`pdf_mode` and `pdf_header` )

- parsing from a directory without a extension passed will now raise an error

- parsing from a directory and extension is `.pdf` will raise an error, since its not supported currently

- changed from tabula to camelot for parsing pdf's

- custom code for header/index columns from parsed tables

**Technical details**

- added the [camelot-py](https://github.com/camelot-dev/camelot) dependency

### CODE REVIEW

**Comment by Imipenem** [comment]
Will fix workflows when done with refactoring, just wanted to open this PR and hear your opinion @Zethson .

File: ehrapy/api/io/_utility_io.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 from pathlib import Path, PurePath
 from typing import Union
 
-supported_extensions = {"csv", "tsv", "h5ad"}
+supported_extensions = {"csv", "tsv", "h5ad", "pdf"}
+# file extensions with formats, that can contain multiple tables/data in one file
+multi_data_extensions = {".pdf"}
```
Comment (Zethson) [inline ehrapy/api/io/_utility_io.py:6]: 1. For consistency `supported_extensions` should then maybe be renamed to single_data_extensions?
2. Using `pdf` above and `.pdf`below. This should be consistent.

File: ehrapy/api/io/read.py [hunk 2]
```diff
@@ -110,35 +113,28 @@ def _read(
...
-            print(
-                "[bold red]Attempted download of missing dataset file(s) failed. Please file an issue at our repository "
-                "[blue]https://github.com/theislab/ehrapy!"
-            )
-        path_cache_dir = settings.cachedir / filename
+        DataReader._check_files_present(filename, backup_url)
+
+        # multi data format extensions like pdf can contain multiple datasets in one single file and are therefore handled as directories when caching
+        path_cache_dir = settings.cachedir / (
+            filename if filename.suffix not in multi_data_extensions else filename.stem
```
Comment (Zethson) [inline ehrapy/api/io/read.py:126]: Wasn't aware of the `stem`function of Pathlib. Cool!

File: ehrapy/api/io/read.py [hunk 3]
```diff
@@ -162,15 +158,64 @@ def _read(
                 raw_anndata, columns_obs_only = DataReader.read_csv(
                     filename, delimiter, index_column, columns_obs_only, cache  # type: ignore
                 )
+                # cache results if wanted
```
Comment (Zethson) [inline ehrapy/api/io/read.py:161]:
```suggestion
                # cache results if desired
```

File: ehrapy/api/io/read.py [hunk 4]
```diff
@@ -162,15 +158,64 @@ def _read(
...
+                if cache:
+                    if not path_cache.parent.is_dir():
+                        path_cache.parent.mkdir(parents=True)
+                    return DataReader._write_cache(raw_anndata, path_cache, columns_obs_only)  # type: ignore
+
+            # read from pdf
```
Comment (Zethson) [inline ehrapy/api/io/read.py:167]: Theoretically a useless comment since the next line already states that this section is about PDF reading :)

File: ehrapy/api/io/read.py [hunk 5]
```diff
@@ -162,15 +158,64 @@ def _read(
...
+        return_mudata: bool = False,
+        cache: bool = False,
+        path_cache_dir: Optional[Path] = None,
+    ) -> Dict[str, AnnData]:
+        """Parse AnnData objects from a directory containing the data files"""
+
+        if not extension:
+            raise ExtensionMissingError(
+                "Reading from directory, but no extension has been provided!. Please "
+                "provide an extension for ehrapy to determine, which file format to read!"
```
Comment (Zethson) [inline ehrapy/api/io/read.py:200]: Can we include the list of supported extensions here? 

->

provide one of XXX extensions to blabal

File: ehrapy/api/io/read.py [hunk 6]
```diff
@@ -264,33 +309,103 @@ def read_csv(
...
+    @staticmethod
+    def read_pdf(
+        filename: Union[Path, Iterator[str]],
+        index_column: Optional[Dict[str, str]] = None,
+        columns_obs_only: Dict[str, Optional[List[Union[str]]]] = None,
+        cache: bool = False,
+        **kwargs,
+    ) -> Tuple[Dict[str, AnnData], Dict[str, Optional[List[Union[str]]]]]:
+        """Read `.pdf`. Since a single pdf can contain multiple tables, those will be read into a dict,
+        like it's done for multiple .csv/.tsv files. Currently, ehrapy only supports parsing single pdf's.
```
Comment (Zethson) [inline ehrapy/api/io/read.py:325]:
```suggestion
        like it's done for multiple .csv/.tsv files. Currently, ehrapy only supports parsing single pdfs.
```

File: ehrapy/api/io/read.py [hunk 7]
```diff
@@ -264,33 +309,103 @@ def read_csv(
...
+            It's also important to note, that this has to be considered when passing "columns_obs_only":
+                .. code-block:: python
+                               import ehrapy.ap as ep
+                               # read pdf
+                               adata_dict = ep.io.read("my_tables.pdf", columns_obs_only={"0":["col1ofTable1", "col2OfTable1"],
+                               "1": ["colOfTable2"]})
+                               # this will put col1 and col2 of Table 0 of my_tables.pdf into obs only for this AnnData object
+                               # and col1 of Table 1 of my_tables.pdf into obs only for this respective AnnData object
+
+            Seems complicating at first glance, but this will allow the most flexibility for users, once one gets used to it.
```
Comment (Zethson) [inline ehrapy/api/io/read.py:355]:
```suggestion
            Seems complicated at first glance, but this will allow the most flexibility for users.
```

File: ehrapy/api/io/read.py [hunk 8]
```diff
@@ -264,33 +309,103 @@ def read_csv(
...
+        # read pandas dataframe
+        no_guess = kwargs.get("guess")
+        # passing guess and set it to False may improve parsed result in cases where other methods will fail
+        initial_df_list = tabula.read_pdf(filename, pages="all", guess=False if no_guess == "no" else True)
+        # in case the index column is misspelled or does not exist
+        if not initial_df_list:
+            raise PdfParsingError(
+                f"Failed parsing file {filename}. Could not parse any data."
+                f"Consider converting your data files to .csv/.tsv before parsing or pass the "
+                f"guess to no parameter to the read function, which may improve parsed results!"
```
Comment (Zethson) [inline ehrapy/api/io/read.py:376]:
```suggestion
                f"guess parameter to the read function, which may improve parsed results!"
```

File: ehrapy/api/io/read.py [hunk 9]
```diff
@@ -264,33 +309,103 @@ def read_csv(
...
+            )
+
+        ann_data_objects = {}
+
+        # one pdf can contain multiple tables, so each of those tables will be one
```
Comment (Zethson) [inline ehrapy/api/io/read.py:381]:
```suggestion
        # one pdf can contain multiple tables, so each of those tables will be one AnnData object
```

File: ehrapy/api/io/read.py [hunk 10]
```diff
@@ -264,33 +309,103 @@ def read_csv(
...
+        for idx, df in enumerate(initial_df_list):
+            print(df)
```
Comment (Zethson) [inline ehrapy/api/io/read.py:383]:
```suggestion
```

**Review by Zethson — COMMENTED** [review]
A couple of minor comments, but overall looks good. 
This PR does unfortunately increase the complexity of the data parsing quite a bit because we handle the case of multiple tables in a single PDF.

Can we please get a couple of tests?

File: ehrapy/api/io/_utility_io.py (hunk 1 above)
Comment (Imipenem) [reply ehrapy/api/io/_utility_io.py:6]: > 1. For consistency supported_extensions should then maybe be renamed to single_data_extensions?

Don't think so, since this variable still holds references for EVERY valid extension, regardless of whether it's multi data or not.

> Using pdf above and .pdf below. This should be consistent.

Will fix

File: ehrapy/api/io/read.py (hunk 2 above)
Comment (Imipenem) [reply ehrapy/api/io/read.py:126]: `Pathlib` has some really cool stuff when it comes to work with filenames etc., yeah

**Comment by Zethson** [comment]
```
      except TypeError as exc:  # suppress unnecessarily long tracebacks
>                   raise TypeError(*exc.args) from None
E                   TypeError: type of argument "backup_url" must be str; got NoneType instead
```

**Comment by Zethson** [comment]
```
    def test_read_pdf(self):
>       ann_data = DataReader.read(dataset_path=f"{_TEST_PATH}/test_pdf.pdf")["test_pdf_0"]

tests/api/io/test_read.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.nox/typeguard-3-8/lib/python3.8/site-packages/typeguard/__init__.py:1033: in wrapper
    retval = func(*args, **kwargs)
ehrapy/api/io/read.py:94: in read
    raw_object = DataReader._read(
.nox/typeguard-3-8/lib/python3.8/site-packages/typeguard/__init__.py:1033: in wrapper
    retval = func(*args, **kwargs)
ehrapy/api/io/read.py:168: in _read
    raw_anndata, columns_obs_only = DataReader.read_pdf(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (PosixPath('/home/zeth/PycharmProjects/ehrapy/tests/api/io/test_data_io/test_pdf.pdf'), None, None, False), kwargs = {}, memo = <typeguard._CallMemo object at 0x7f6fd7ea4580>
retval = ({'test_pdf_0': AnnData object with n_obs × n_vars = 31 × 10
    layers: 'original'}, None)

    def wrapper(*args, **kwargs):
        memo = _CallMemo(python_func, _localns, args=args, kwargs=kwargs)
        check_argument_types(memo)
        retval = func(*args, **kwargs)
        try:
            check_return_type(retval, memo)
        except TypeError as exc:
>           raise TypeError(*exc.args) from None
E           TypeError: type of the return value[1] must be a dict; got NoneType instead

.nox/typeguard-3-8/lib/python3.8/site-packages/typeguard/__init__.py:1037: TypeError
```

@Imipenem did you try this already: https://stackoverflow.com/questions/59224551/data-missing-while-reading-pdf-file-using-tabula-and-python ?

**Comment by Zethson** [comment]
@Imipenem should we check for an available JDK before attempting to run tabulapy?

**Comment by Imipenem** [comment]
>@Imipenem did you try this already: https://stackoverflow.com/questions/59224551/data-missing-while-reading-pdf-file-using-tabula-and-python ?

Yes but this is not useful as a general solution, some pdf files might get better parsed when using stream mode others using lattice mode or when one set `guess` parameter to False.

We can implement all of them into one function and if the user is not satisfied with the parsing result, one could try another mode etc. 


But unfortunately this won't solve the issue completely and make it robust.

**Comment by Zethson** [comment]
> > @Imipenem did you try this already: [https://stackoverflow.com/questions/59224551/data-missing-while-reading-pdf-file-using-tabula-and-python ](https://stackoverflow.com/questions/59224551/data-missing-while-reading-pdf-file-using-tabula-and-python%C2%A0)?
> 
> Yes but this is not useful as a general solution, some pdf files might get better parsed when using stream mode others using lattice mode or when one set `guess` parameter to False.
> 
> We can implement all of them into one function and if the user is not satisfied with the parsing result, one could try another mode etc.
> 
> But unfortunately this won't solve the issue completely and make it robust.

All right. Let's properly document this then. All we can do.

But yeah, please implement a quick JDK check.

**Comment by Imipenem** [comment]
See https://github.com/camelot-dev/camelot/issues/278.

As discussed: Better parsing results ( no missing data) than tabula-py, but a mess to deal with row (index) names or column (header) names.

File: ehrapy/api/io/read.py [hunk 11]
```diff
@@ -162,15 +158,64 @@ def _read(
...
+        filename: Path,
+        extension: Optional[str] = None,
+        delimiter: Optional[str] = None,
+        index_column: Optional[Union[Dict[str, Union[str, int]], Union[str, int]]] = None,
+        columns_obs_only: Optional[Union[Dict[str, List[str]], List[str]]] = None,
+        return_mudata: bool = False,
+        cache: bool = False,
+        path_cache_dir: Optional[Path] = None,
+    ) -> Dict[str, AnnData]:
+        """Parse AnnData objects from a directory containing the data files"""
```
Comment (Zethson) [inline ehrapy/api/io/read.py:194]: Does this warrant a full docstring?

File: ehrapy/api/io/read.py [hunk 12]
```diff
@@ -264,33 +309,121 @@ def read_csv(
...
+                           # "my_tables_3": AnnData object with C obs, D vars.
+                           # }
+
+            The example above illustrates, that the different tables will be stored under the index of the
+            order they appeared in the pdf, prefixed by the pdf filename. This ensures uniqueness, especially in cases
+            when multiple pdfs will be read, with multiple tables per file.
+
+            It's also important to note, that this has to be considered when passing "columns_obs_only":
+                .. code-block:: python
+                               import ehrapy.ap as ep
```
Comment (Zethson) [inline ehrapy/api/io/read.py:348]:
```suggestion
                               import ehrapy.api as ep
```

File: ehrapy/api/io/read.py [hunk 13]
```diff
@@ -264,33 +309,121 @@ def read_csv(
...
+        **kwargs,
+    ) -> Tuple[Dict[str, AnnData], Optional[Dict[str, Optional[List[Union[str]]]]]]:
+        """Read `.pdf`. Since a single pdf can contain multiple tables, those will be read into a dict,
+        like it's done for multiple .csv/.tsv files. Currently, ehrapy only supports parsing single pdfs.
+
+        Consider the following example: "my_tables.pdf" contains three different tables, which may
+        also differ in size.
+
+            .. code-block:: python
+                           import ehrapy.ap as ep
```
Comment (Zethson) [inline ehrapy/api/io/read.py:331]:
```suggestion
                           import ehrapy.api as ep
```

File: ehrapy/api/io/read.py [hunk 14]
```diff
@@ -264,33 +309,121 @@ def read_csv(
...
+        ann_data_objects = {}
+
+        # one pdf can contain multiple tables, so each of those tables will be one AnnData object
+        for idx, table in enumerate(initial_df_list):
+            df = table.df
+            is_set = isinstance(header, set)
+            # defaults to True, so if header has not been set, assume first row is column names row
+            # if header is a set and table number idx is in header, also assume first row is also column names row
+            if header is None or (is_set and idx in header):
+                # when the entry in top left corner is empty assume, table has header and index names stored in first row/first column
```
Comment (Zethson) [inline ehrapy/api/io/read.py:391]: I would add a yellow print as a warning when we do this.

File: tests/api/io/test_read.py [hunk 15]
```diff
@@ -32,58 +32,74 @@ def test_read_tsv(self):
...
-        assert id(ann_data.layers["original"]) != id(ann_data.X)
-        assert list(ann_data.obs.index) == ["0", "1", "2", "3", "4"]
+        assert adata.X.shape == (5, 4)
+        assert (adata.X == matrix).all()
+        assert adata.var_names.to_list() == ["clinic_id", "los_days", "b12_values", "survival"]
+        assert (adata.layers["original"] == matrix).all()
+        assert id(adata.layers["original"]) != id(adata.X)
+        assert list(adata.obs.index) == ["0", "1", "2", "3", "4"]
+
+    def test_read_pdf(self):
```
Comment (Zethson) [inline tests/api/io/test_read.py:53]: Can we please add one test for a PDF with a proper header and one where the top left is empty?

**Review by Zethson — CHANGES_REQUESTED** [review]
A few too many `#type: ignore`. What is the issue here?

File: ehrapy/api/io/read.py (hunk 14 above)
Comment (Imipenem) [reply ehrapy/api/io/read.py:391]: I'd say no. Imagine a user parses like 20 tables: Then he will get 20 print warnings since this should then print for all tables, since those could behave differently.
